test/test_cuda.py

```diff
@@ -1069,8 +1069,13 @@ def test_mul(dtype):
 
             self.assertEqual(x * y, 4.5)
             self.assertEqual(y * x, 4.5)
-            with self.assertRaisesRegex(RuntimeError, "doesn't match the desired"):
-                y *= x
+
+            # *= is an in-place operation equivalent to prod_(). Doesn't modify
+            # dtype, instead casts x to int.
+            y *= x
+            self.assertEqual(y, 3)
```

Comment:
I thought you and @colesbury said this should error?

In any case, I don't believe the comment or behavior is correct, even if we were to allow this.  See:

```
>>> x=np.array([2,3])

>>> np.multiply(x, 1.5, out=x, casting='unsafe')
array([3, 4])
```

That's not casting x to an int.

Reply:
`y *= x` is equivalent to `np.multiply(y, 1.5, out=y, casting='unsafe')` where `x` is the `1.5`. The 1.5 gets cast to an int / the dtype of y doesn't change.

Yes @colesbury is saying this should error. I thought it shouldn't. It currently doesn't, neither in this PR nor on master.

Reply:
They're not the same. NumPy casts the result of (y * 1.5) to the to an int. The legacy PyTorch behavior casts (1.5) to an int. @gchanan's example above shows the difference.

Reply:
I just meant `np.multiply(x, 1.5, out=x, casting='unsafe')` is equivalent to `x *= 1.5`, not the `y *= x` from the unit test. 

That's correct that part being cast to int is either the 1.5 (in pytorch) or the result of `tensor*1.5` (in numpy), not the tensor itself. It seems that an implementation detail in python prevents numpy from actually re-using the same memory of the inplace tensor when performing in-place operations. We have the ability to (continue to) perform operations directly inplace if we choose.

If we choose to allow it, I believe the comment is correct for this branch. If we cast the result of the `(tensor * 1.5)` operation as in numpy instead of just the `1.5` as we do in master, that could result in creating an extra temporary tensor, and would be a different backwards-incompatible change. I don't think that's the behavior we would want.